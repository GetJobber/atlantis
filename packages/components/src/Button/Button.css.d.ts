export const button: string;
export const work: string;
export const learning: string;
export const destructive: string;
export const cancel: string;
export const subtle: string;
export const primary: string;
export const secondary: string;
export const tertiary: string;
export const disabled: string;
export const small: string;
export const base: string;
export const large: string;
export const onlyIcon: string;
export const hasIconAndLabel: string;
export const iconOnRight: string;
export const fullWidth: string;
export const loading: string;
export const animateStripes: string;
