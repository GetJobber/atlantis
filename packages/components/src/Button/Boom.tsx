import React, { useEffect, useState } from "react";
import { Particle } from "./Particle";
import { OutParticle } from "./OutParticle";

export const Boom = ({
  enabled,
  particles = 50,
  boomType = "firework",
  scheme = "all",
}: {
  readonly enabled: boolean;
  readonly particles: number;
  readonly boomType?: string;
  readonly scheme?: string;
}) => {
  const [particleAmount] = useState(particles);
  const [particlesEnabled, setParticlesEnabled] = useState(false);
  const [loopArray] = useState(Array.from({ length: particleAmount }));
  useEffect(() => {
    if (enabled) {
      requestAnimationFrame(() => {
        setParticlesEnabled(true);
      });
    } else {
      setParticlesEnabled(false);
    }
  }, [enabled]);

  return (
    <>
      {enabled &&
        loopArray.map((_, index) => {
          console.log("BOOMT!", boomType);

          return boomType === "firework" ? (
            <Particle key={index} enabled={particlesEnabled} scheme={scheme} />
          ) : (
            <OutParticle
              key={index}
              enabled={particlesEnabled}
              scheme={scheme}
            />
          );
        })}
    </>
  );
};
