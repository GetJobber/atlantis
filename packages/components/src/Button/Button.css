:root {
  --button--color-work: var(--color-green);
  --button--color-work--hover: var(--color-green--dark);

  --button--color-learning: var(--color-lightBlue);
  --button--color-learning--hover: var(--color-lightBlue--dark);

  --button--color-destructive: var(--color-red);
  --button--color-destructive--hover: var(--color-red--dark);

  --button--color-cancel: var(--color-greyBlue);
  --button--color-cancel--hover: var(--color-grey--lightest);

  --button--bg-color-cancel: var(--color-white);
  --button--color-disabled: var(--color-grey--lighter);
  --button--color-secondary: var(--color-white);
  --button--color-tertiary: var(--color-white);

  --button--duration-loading: 2s;
}

.button {
  display: inline-flex;
  min-height: calc(var(--base-unit) * 2.25);
  box-sizing: border-box;
  margin: 0;
  border: 1px solid transparent;
  border-radius: var(--radius-base);
  color: transparent;
  -webkit-font-smoothing: antialiased;
  text-align: center;
  text-decoration: none;
  vertical-align: middle;
  background-color: transparent;
  cursor: pointer;
  transition: all var(--timing-base);
  appearance: none;
  align-items: center;
  justify-content: center;
}

.button:focus,
a.button:focus {
  outline-color: var(--color-focus);
}

/* Variation */

.work {
  border-color: var(--button--color-work);
  color: var(--button--color-work);
  background-color: var(--button--color-work);
}

.work:hover,
a.work:hover,
.work:focus,
a.work:focus {
  border-color: var(--button--color-work--hover);
  background-color: var(--button--color-work--hover);
}

.learning {
  border-color: var(--button--color-learning);
  background-color: var(--button--color-learning);
}

.learning:hover,
a.learning:hover,
.learning:focus,
a.learning:focus {
  border-color: var(--button--color-learning--hover);
  background-color: var(--button--color-learning--hover);
}

.destructive {
  border-color: var(--button--color-destructive);
  background-color: var(--button--color-destructive);
}

.destructive:hover,
a.destructive:hover,
.destructive:focus,
a.destructive:focus {
  border-color: var(--button--color-destructive--hover);
  background-color: var(--button--color-destructive--hover);
}

/* Cancel is special because, by default, it's styled as a secondary button */

.cancel {
  border-color: var(--button--color-cancel);
  background-color: var(--button--bg-color-cancel);
}

.cancel:focus,
.cancel:hover,
a.cancel:focus,
a.cancel:hover {
  background-color: var(--button--color-cancel--hover);
  filter: brightness(80%);
}

/* Type */

.primary {
  color: var(--color-white);
}

.secondary {
  background-color: var(--button--color-secondary);
}

.secondary:hover,
.secondary:focus,
a.secondary:hover,
a.secondary:focus {
  background-color: var(--button--color-secondary);
  filter: brightness(80%);
}

.tertiary {
  border-color: var(--button--color-tertiary);
  background-color: var(--button--color-tertiary);
}

.tertiary:hover,
.tertiary:focus,
a.tertiary:hover,
a.tertiary:focus {
  border-color: transparent;
  background-color: var(--color-grey--lightest);
}

/* Disabled */

.disabled,
.disabled:hover {
  border-color: var(--button--color-disabled);
  user-select: none;
  background-color: var(--button--color-disabled);
  pointer-events: none;
  cursor: not-allowed;
}

/* Sizes */

.small {
  min-height: calc(var(--base-unit) * 1.5);
  padding: 0 var(--space-small);
}

.base {
  min-height: calc(var(--base-unit) * 2.25);
  padding: 0 calc(var(--space-base) - 4px);
}

.large {
  min-height: calc(var(--base-unit) * 3);
  padding: 0 var(--space-base);
}

/* Icon */

.onlyIcon {
  width: calc(var(--base-unit) * 2.25);
  padding: 0;
}

.onlyIcon.small {
  width: calc(var(--space-small) * 3);
}

.onlyIcon.base {
  width: calc(var(--space-base) * 2.25);
}

.onlyIcon.large {
  width: calc(var(--space-large) * 2);
}

.hasIconAndLabel {
  width: auto;
}

.hasIconAndLabel > *:first-child {
  margin-right: var(--space-small);
}

.iconOnRight > *:first-child {
  margin-left: var(--space-small);
  margin-right: 0;
  order: 1;
}

.fullWidth {
  width: 100%;
}

.loading {
  background-image: linear-gradient(
    135deg,
    rgba(0, 0, 0, 10%) 0%,
    rgba(0, 0, 0, 10%) 25%,
    transparent 25%,
    transparent 50%,
    rgba(0, 0, 0, 10%) 50%,
    rgba(0, 0, 0, 10%) 75%,
    transparent 75%,
    transparent 100%
  );
  background-size: var(--space-larger) var(--space-larger);
  pointer-events: none;
  cursor: not-allowed;
  animation-duration: var(--button--duration-loading);
  animation-iteration-count: infinite;
  animation-name: animateStripes;
  animation-timing-function: linear;
}

.primary.loading:not(.cancel) {
  background-image: linear-gradient(
    135deg,
    rgba(255, 255, 255, 25%) 0%,
    rgba(255, 255, 255, 25%) 25%,
    transparent 25%,
    transparent 50%,
    rgba(255, 255, 255, 25%) 50%,
    rgba(255, 255, 255, 25%) 75%,
    transparent 75%,
    transparent 100%
  );
}

@keyframes animateStripes {
  0% {
    background-position: 0 0;
  }
  100% {
    background-position: var(--space-larger) var(--space-larger);
  }
}
