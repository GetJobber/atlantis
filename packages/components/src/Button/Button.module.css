:root {
  --public--button--color-variation: var(--color-interactive);
  --public--button--color-variation--hover: var(--color-interactive--hover);
}

.button {
  --button--color-surface: var(--color-surface);
  --button--color-surface--hover: var(--color-surface--hover);
  --button--color-primaryLabel: var(--color-surface);
  --button--duration-loading: var(--timing-loading--extended);
  --button--size: 40px;

  display: inline-flex;
  min-height: var(--button--size);
  box-sizing: border-box;
  margin: 0;
  border: var(--border-base) solid var(--public--button--color-variation);
  border-radius: var(--radius-base);
  color: var(--public--button--color-variation);
  -webkit-font-smoothing: antialiased;
  line-height: 1;
  text-align: center;
  text-decoration: none;
  vertical-align: middle;
  background-color: var(--button--color-surface);
  cursor: pointer;
  transition: all var(--timing-base) ease-out;
  appearance: none;
  fill: var(--public--button--color-variation);
  align-items: center;
  justify-content: center;
  gap: var(--space-smaller);
}

.button * {
  color: inherit;
  fill: inherit !important;
  /* required to over-ride very specific fill from <Icon> */
}

.button:hover,
a.button:hover,
.button:focus-visible,
a.button:focus-visible,
.button:active,
a.button:active {
  border-color: var(--public--button--color-variation--hover);
  color: var(--public--button--color-variation--hover);
  fill: var(--public--button--color-variation--hover);
  background-color: var(--button--color-surface--hover);
}

.button:focus-visible,
a.button:focus-visible,
.button:active,
a.button:active {
  box-shadow: var(--shadow-focus);
  outline: transparent;
}

/* Variation */

.work {
  --public--button--color-variation: var(--color-interactive);
  --public--button--color-variation--hover: var(--color-interactive--hover);
}

.learning,
.subtle {
  --public--button--color-variation: var(--color-interactive--subtle);
  --public--button--color-variation--hover: var(
    --color-interactive--subtle--hover
  );
}

.destructive {
  --public--button--color-variation: var(--color-destructive);
  --public--button--color-variation--hover: var(--color-destructive--hover);
}

/* Type */

.primary:not(.subtle, .disabled),
.primary:not(.subtle, .disabled):hover,
.primary:not(.subtle, .disabled):focus-visible {
  color: var(--button--color-primaryLabel);
  background-color: var(--public--button--color-variation);
  fill: var(--button--color-primaryLabel);
}

.primary:not(.subtle, .disabled):hover,
.primary:not(.subtle, .disabled):focus-visible {
  background-color: var(--public--button--color-variation--hover);
}

.subtle.primary {
  color: var(--public--button--color-variation);
  fill: var(--public--button--color-variation);
  background-color: var(--button--color-surface);
}

.secondary,
.subtle.primary {
  border-color: var(--color-border--interactive);
}

.subtle.primary:hover,
.subtle.primary:focus-visible {
  color: var(--public--button--color-variation--hover);
  background-color: var(--button--color-surface--hover);
  fill: var(--public--button--color-variation--hover);
}

.secondary:hover,
.subtle.primary:hover,
.secondary:focus-visible,
.subtle.primary:focus-visible {
  border-color: var(--public--button--color-variation--hover);
}

.subtle.secondary {
  background-color: var(--button--color-surface);
}

.tertiary,
.subtle.secondary {
  border-color: transparent;
}

.tertiary {
  border-color: transparent;
  text-decoration: underline;
  text-decoration-thickness: var(--border-thick);
  text-underline-offset: var(--space-smallest);
}

.subtle.secondary:hover,
.subtle.secondary:focus-visible {
  background-color: var(--button--color-surface--hover);
}

.subtle.secondary:hover,
.subtle.secondary:focus-visible,
a.subtle.secondary:hover,
a.subtle.secondary:focus-visible,
.tertiary:hover,
.tertiary:focus-visible,
a.tertiary:hover,
a.tertiary:focus-visible {
  border-color: transparent;
}

.subtle.tertiary {
  border-color: transparent;
  background-color: transparent;
}

.subtle.tertiary:hover,
.subtle.tertiary:focus-visible,
a.subtle.tertiary:hover,
a.subtle.tertiary:focus-visible {
  color: var(--color-interactive--subtle--hover);
  background-color: var(--color-surface--hover);
  fill: var(--color-interactive--subtle--hover);
}

/* Disabled */

.button.disabled,
a.button.disabled,
.button.disabled:hover,
a.button.disabled:hover {
  border-color: var(--color-disabled--secondary);
  color: var(--color-disabled);
  user-select: none;
  background-color: var(--color-disabled--secondary);
  fill: var(--color-disabled);
  pointer-events: none;
  cursor: not-allowed;
}

/* Sizes */

.small {
  --button--size: 32px;
  padding: var(--space-smaller) calc(var(--space-base) - var(--space-smaller));
}

.base {
  padding: 0 var(--space-base);
}

.large {
  --button--size: 48px;
  padding: 0 calc(var(--space-base) + var(--space-smaller));
}

/* Icon */

.button [data-button-icon] {
  flex-shrink: 0;
}

.buttonChildren:has(> [data-button-icon]:only-child) {
  width: var(--button--size);
  padding: 0;
}

/** Icon on Left */
.buttonChildren
  > [data-button-icon]:first-of-type:first-child:not(:only-child):not(
    :last-child
  ) {
  margin-left: calc(var(--space-smaller) * -1);
}

/** Icon on Right */
.buttonChildren
  > [data-button-icon]:last-of-type:last-child:not(:only-child):not(
    :first-child
  ) {
  margin-right: calc(var(--space-smaller) * -1);
}

.fullWidth {
  width: 100%;
}

.loading {
  background-image: linear-gradient(
    135deg,
    rgba(0, 0, 0, 0.1) 0%,
    rgba(0, 0, 0, 0.1) 25%,
    transparent 25%,
    transparent 50%,
    rgba(0, 0, 0, 0.1) 50%,
    rgba(0, 0, 0, 0.1) 75%,
    transparent 75%,
    transparent 100%
  );
  background-size: var(--space-larger) var(--space-larger);
  pointer-events: none;
  cursor: not-allowed;
  animation-duration: var(--button--duration-loading);
  animation-iteration-count: infinite;
  animation-name: animateStripes;
  animation-timing-function: linear;
}

.primary.loading:not(.subtle) {
  background-image: linear-gradient(
    135deg,
    rgba(255, 255, 255, 0.25) 0%,
    rgba(255, 255, 255, 0.25) 25%,
    transparent 25%,
    transparent 50%,
    rgba(255, 255, 255, 0.25) 50%,
    rgba(255, 255, 255, 0.25) 75%,
    transparent 75%,
    transparent 100%
  );
}

@keyframes animateStripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: var(--space-larger) var(--space-larger);
  }
}
