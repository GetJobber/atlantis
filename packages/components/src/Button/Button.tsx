import React from "react";
import classnames from "classnames";
import { XOR } from "ts-xor";
import { Link } from "react-router-dom";
import { IconNames } from "@jobber/design";
import styles from "./Button.css";
import { Typography } from "../Typography";
import { Icon } from "../Icon";

type ButtonType = "button" | "submit";

interface ButtonFoundationProps {
  /**
   * Used for screen readers. Will override label on screen
   * reader if present.
   */
  readonly ariaLabel?: string;
  readonly ariaControls?: string;
  readonly ariaHaspopup?: boolean;
  readonly ariaExpanded?: boolean;
  readonly disabled?: boolean;
  readonly external?: boolean;
  readonly fullWidth?: boolean;
  readonly icon?: IconNames;
  readonly iconOnRight?: boolean;
  readonly id?: string;
  readonly label?: string;
  readonly loading?: boolean;
  readonly size?: "small" | "base" | "large";
  onClick?(): void;
}

interface ButtonIconProps extends ButtonFoundationProps {
  readonly icon: IconNames;
  readonly ariaLabel: string;
}

interface ButtonLabelProps extends ButtonFoundationProps {
  readonly label: string;
}

interface ButtonAnchorProps extends ButtonFoundationProps {
  /**
   * Used to create an 'href' on an anchor tag.
   */
  readonly url?: string;
}

interface ButtonLinkProps extends ButtonFoundationProps {
  /**
   * Used for client side routing. Only use when inside a routed component.
   */
  readonly to?: string;
}

interface BaseActionProps extends ButtonFoundationProps {
  readonly variation?: "work" | "learning";
  readonly type?: "primary" | "secondary" | "tertiary";
}

interface DestructiveActionProps extends ButtonFoundationProps {
  readonly variation: "destructive";
  readonly type?: "primary" | "secondary" | "tertiary";
}

interface CancelActionProps extends ButtonFoundationProps {
  readonly variation: "cancel";
  readonly type?: "secondary" | "tertiary";
}

interface SubmitActionProps
  extends Omit<ButtonFoundationProps, "external" | "onClick"> {
  readonly variation?: "work";
  readonly type?: "primary";
  readonly submit: boolean;
}

interface SubmitButtonProps
  extends Omit<ButtonFoundationProps, "external" | "onClick"> {
  /**
   * Allows the button to submit a form
   */
  submit: boolean;
}

export type ButtonProps = XOR<
  BaseActionProps,
  XOR<DestructiveActionProps, XOR<CancelActionProps, SubmitActionProps>>
> &
  XOR<SubmitButtonProps, XOR<ButtonLinkProps, ButtonAnchorProps>> &
  XOR<ButtonIconProps, ButtonLabelProps>;

export function Button(props: ButtonProps) {
  const {
    ariaControls,
    ariaHaspopup,
    ariaExpanded,
    ariaLabel,
    disabled = false,
    external,
    fullWidth,
    icon,
    label,
    iconOnRight,
    id,
    loading,
    onClick,
    size = "base",
    type = "primary",
    url,
    to,
    variation = "work",
    submit,
  } = props;

  const buttonClassNames = classnames(styles.button, styles[size], {
    [styles.onlyIcon]: icon && !label,
    [styles.hasIconAndLabel]: icon && label,
    [styles.iconOnRight]: iconOnRight,
    [styles[variation]]: variation,
    [styles[type]]: type,
    [styles.disabled]: disabled,
    [styles.fullWidth]: fullWidth,
    [styles.loading]: loading,
  });

  const buttonType: ButtonType = submit ? "submit" : "button";

  const tagProps = {
    className: buttonClassNames,
    disabled,
    id,
    ...(!disabled && { href: url }),
    ...(!disabled && { onClick: onClick }),
    ...(external && { target: "_blank" }),
    ...(url === undefined && to === undefined && { type: buttonType }),
    "aria-controls": ariaControls,
    "aria-haspopup": ariaHaspopup,
    "aria-expanded": ariaExpanded,
    "aria-label": ariaLabel,
  };

  const buttonInternals = <ButtonInternals {...props} />;

  if (to) {
    return (
      <Link {...tagProps} to={to}>
        {buttonInternals}
      </Link>
    );
  }

  const Tag = url ? "a" : "button";

  return <Tag {...tagProps}>{buttonInternals}</Tag>;
}

function ButtonInternals({
  label,
  icon,
  variation = "work",
  type = "primary",
  disabled,
  size = "base",
}: ButtonProps) {
  return (
    <>
      {icon && (
        <Icon
          name={icon}
          size={size}
          color={getColor(variation, type, disabled)}
        />
      )}
      <Typography
        element="span"
        textCase="uppercase"
        fontWeight="extraBold"
        size={getTypeSizes(size)}
        textColor={getColor(variation, type, disabled)}
      >
        {label}
      </Typography>
    </>
  );
}

function getTypeSizes(size: string) {
  switch (size) {
    case "small":
      return "smaller";
    case "large":
      return "base";
    default:
      return "small";
  }
}

function getColor(variation: string, type: string, disabled?: boolean) {
  if (type === "primary" && variation !== "cancel" && !disabled) {
    return "white";
  }

  if (disabled) {
    return "grey";
  }

  switch (variation) {
    case "learning":
      return "lightBlue";
    case "destructive":
      return "red";
    case "cancel":
      return "greyBlue";
    default:
      return "green";
  }
}
