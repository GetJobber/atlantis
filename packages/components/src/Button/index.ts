export { Button } from "./Button";
export * from "./Button.types";
