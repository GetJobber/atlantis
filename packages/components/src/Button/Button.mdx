---
name: Button
menu: Components
route: /components/button
showDirectoryLink: true
---

import { Playground, Props } from "docz";
import { ComponentStatus } from "@jobber/docx";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { Form } from "@jobber/components/Form";
import { InputNumber } from "@jobber/components/InputNumber";
import { Content } from "@jobber/components/Content";
import { Button } from "@jobber/components/Button";

# Button

<ComponentStatus stage="ready" responsive="yes" accessible="yes" />

Buttons are a core user interface component, as they allow users to initiate,
complete, and reverse actions.

```ts
import { Button } from "@jobber/components/Button";
```

<Playground>
  <Button
    label="Do the thing"
    onClick={() => {
      alert("âœ…");
    }}
  />
</Playground>

## Props

<Props of={Button} />

---

## Design & Usage Guidelines

The right button to use depends on:

- the type of action your user is attempting to complete
- the hierarchy of the interface the button lives in

### Work actions

Use Work Actions buttons to enable user interactions in the goal of completing
work in Jobber.

#### Primary

The most important action in a view. Examples include the Save Job button, Log
In, or the main action in a dialog. There should be one Primary action per view,
at most.

<Playground>
  <Button label="Do the thing" />
</Playground>

#### Secondary

A secondary button should be used when...

- there is one or more alternative, but less important, actions to the primary
  action

<Playground>
  <Button label="Do the thing" type="secondary" />
</Playground>

#### Tertiary

A tertiary button should be used to complete actions within the view such as
editing, revealing details, or navigating within the view.

<Playground>
  <Button label="Do the thing" type="tertiary" />
</Playground>

### Learning Actions

Use the Learning Actions buttons to enable user interactions in the goal of
learning about Jobber, whether for marketing or coaching purposes.

#### Primary, Secondary, Tertiary

The usage pattern for Learning Actions should follow the same logic as Work
Actions for all three levels.

<Playground>
  <Button
    label="Learn More"
    variation="learning"
    url="//getjobber.com"
    external={true}
  />
</Playground>

<Playground>
  <Button
    label="Learn More"
    variation="learning"
    type="secondary"
    url="//getjobber.com"
    external={true}
  />
</Playground>

<Playground>
  <Button
    label="Learn More"
    variation="learning"
    type="tertiary"
    url="//getjobber.com"
    external={true}
  />
</Playground>

### Destructive Actions

Destructive actions remove something from Jobber.

#### Primary

A Primary destructive action will permanently destroy an object carrying
signficant data, such as a job, quote, or client. This destructive action should
be contained in a confirmation dialog triggered by clicking a Secondary
destructive button.

#### Secondary or Tertiary

A Secondary or tertiary destructive action will either...

- destroy an item on an object that can be quickly restored, such as a line item
  on a job
- trigger a dialog to confirm a Primary destructive action

<Playground>
  <Button label="Delete" variation="destructive" />
</Playground>

<Playground>
  <Button label="Delete" variation="destructive" type="secondary" />
</Playground>

### Cancel Actions

Cancel actions allow the user to opt out of completing an action they have
triggered, whether that action is Work, Learning, or Destructive.

<Playground>
  <Button label="Cancel" variation="cancel" />
</Playground>

## Disabled

Buttons can be disabled for a number of reasons, users may need to do an action
first or they may not be ready to use that button.

<Playground>
  <Button label="Do the thing" disabled />
</Playground>

## Sizes

Buttons comes in 2 extra sizes.

### Small

<Playground>
  <Button label="Do the thing" size="small" />
</Playground>

### Large

<Playground>
  <Button label="Do the thing" size="large" />
</Playground>

## Icons

You can use icons inside buttons. They can stand-alone, or appear before, after,
or on top of the button label. They can also be added before or after the label.

**Note:** If an icon is used in a stand-alone method, an `ariaLabel` must be
present to describe the button.

<Playground>
  <Button type="secondary" icon="user" ariaLabel="I'm a person" />
  <Button label="More" type="secondary" icon="more" />
  <Button
    label="Actions"
    type="secondary"
    icon="arrowDown"
    iconOnRight={true}
  />
</Playground>

## Loading

A way to communicate to the user that a background action is being performed.

<Playground>
  <Button
    label="Deleting..."
    type="primary"
    variation="destructive"
    loading={true}
  />
  <Button
    label="Loading..."
    type="tertiary"
    variation="learning"
    loading={true}
  />
  <Button label="Canceling..." variation="cancel" loading={true} />
</Playground>

## Client Side Routing

A `Button` can be used for client side routing as well. When using a Button to
handle the client side routing, use the `to` prop. Notice when you click below
that the url will change to the appropriate route.

<Playground>
  {() => {
    return (
      <Router basename="/components/button">
        <Button label="Home" to="/" />
        <Button label="Office" to="/office" />
        <Button label="Dentist" to="/dentist" />
        <hr />
        <Switch>
          <Route exact path="/">
            This is my home, time to get cozy.
          </Route>
          <Route exact path="/office">
            This is my office, time to get to work.
          </Route>
          <Route exact path="/dentist">
            This is the dentist, time to get my teeth fixed.
          </Route>
        </Switch>
      </Router>
    );
  }}
</Playground>

## Form Submit

Passing the `submit` prop will allow a `Button` to submit a
[Form](/components/form). Since submitting a form is a specific action, only
`variation="work"` and `type="primary"` are allowed.

Since this type of `Button` will only be used to submit a form, it does not make
sense to allow the `external`, `onClick`, `to`, or `url` props.

<Playground>
  <Form onSubmit={() => alert("Wow, what a number")}>
    <Content>
      <InputNumber placeholder="Pick a number" />
      <Button label="Submit" submit={true} />
    </Content>
  </Form>
</Playground>
