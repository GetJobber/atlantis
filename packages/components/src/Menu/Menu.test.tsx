import React from "react";
import renderer from "react-test-renderer";
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { Menu } from ".";
import { Button } from "../Button";

afterEach(cleanup);
jest.mock("uuid");

it("renders a Menu", () => {
  const tree = renderer
    .create(
      <Menu
        items={[
          {
            header: "Send as...",
            actions: [
              {
                label: "Text Message",
                icon: "sms",
              },
              {
                label: "Email",
                icon: "email",
              },
            ],
          },
        ]}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders a Menu with custom activator", () => {
  const tree = renderer
    .create(
      <Menu
        activator={<Button label="Menu" />}
        items={[
          {
            header: "Send as...",
            actions: [
              {
                label: "Text Message",
                icon: "sms",
              },
              {
                label: "Email",
                icon: "email",
              },
            ],
          },
        ]}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("it should open and close the menu", async () => {
  const header = "Mark as...";
  const actionLabel = "Awaiting Response";
  const clickHandler = jest.fn();
  const actions = [
    {
      header: header,
      actions: [{ label: actionLabel, onClick: clickHandler }],
    },
  ];

  const { getByRole, queryAllByText } = render(<Menu items={actions} />);

  fireEvent.click(getByRole("button"));
  expect(queryAllByText(actionLabel).length).toBe(1);

  fireEvent.click(getByRole("menuitem"));
  expect(clickHandler).toHaveBeenCalledTimes(1);

  await waitFor(() => {
    expect(queryAllByText(actionLabel).length).toBe(0);
  });
});

test("it should passthrough an activator's click action", () => {
  const clickHandler = jest.fn();
  const actions = [];

  const { getByRole } = render(
    <Menu
      activator={<Button label="Menu" onClick={clickHandler} />}
      items={actions}
    />,
  );

  fireEvent.click(getByRole("button"));
  expect(clickHandler).toHaveBeenCalledTimes(1);
});

it("should focus first action item from the menu when activated", () => {
  const { getByRole } = render(
    <Menu
      activator={<Button label="Menu" />}
      items={[
        {
          header: "Send as...",
          actions: [
            {
              label: "Text Message",
              icon: "sms",
            },
            {
              label: "Email",
              icon: "email",
            },
          ],
        },
      ]}
    />,
  );

  fireEvent.click(getByRole("button"));
  const firstMenuItem = screen.getAllByRole("menuitem")[0];
  expect(firstMenuItem).toHaveFocus();
});
