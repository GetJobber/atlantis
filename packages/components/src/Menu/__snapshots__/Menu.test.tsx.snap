// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Menu 1`] = `
<div
  className="wrapper"
>
  <button
    aria-controls="123e4567-e89b-12d3-a456-426655440002"
    aria-expanded={false}
    aria-haspopup={true}
    className="button base hasIconAndLabel work secondary fullWidth"
    disabled={false}
    id="123e4567-e89b-12d3-a456-426655440001"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="ZDlJU6tECg7ouG-b27gya _3ctOeOMP7zLhU37n25xoZC"
      data-testid="more"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        className=""
        d="M170.667 512c0-47.13 38.205-85.333 85.333-85.333s85.333 38.204 85.333 85.333c0 47.13-38.205 85.333-85.333 85.333s-85.333-38.204-85.333-85.333zM426.667 512c0-47.13 38.204-85.333 85.333-85.333s85.333 38.204 85.333 85.333c0 47.13-38.204 85.333-85.333 85.333s-85.333-38.204-85.333-85.333zM768 426.667c-47.13 0-85.333 38.204-85.333 85.333s38.204 85.333 85.333 85.333c47.13 0 85.333-38.204 85.333-85.333s-38.204-85.333-85.333-85.333z"
      />
    </svg>
    <span
      className="base extraBold small uppercase"
    >
      More Actions
    </span>
  </button>
</div>
`;

exports[`renders a Menu with custom activator 1`] = `
<div
  className="wrapper"
>
  <button
    aria-controls="123e4567-e89b-12d3-a456-426655440004"
    aria-expanded={false}
    aria-haspopup={true}
    className="button base work primary"
    disabled={false}
    id="123e4567-e89b-12d3-a456-426655440003"
    onClick={[Function]}
    type="button"
  >
    <span
      className="base extraBold small uppercase"
    >
      Menu
    </span>
  </button>
</div>
`;
